File: codes/julia/operator_optional_chaining.jl
using Printf
using Statistics
using JSON

println("\n# JavaScript-like Optional Chaining Operator (?.) in Julia")

# There's no JavaScript-like Optional Chaining Operator (?.) in Julia.
# But, we can create our own function to mimic it in Julia.

# JavaScript-like Optional Chaining Operator (?.) function
optional_chaining_v1 = (an_object, object_properties_array...) -> reduce((current_result, current_item) -> ((current_result === nothing) ? get(an_object, current_item, nothing) : get(current_result, current_item, nothing)), object_properties_array; init=nothing)

# JavaScript-like Optional Chaining Operator (?.) function
optional_chaining_v2 = (an_object, object_properties_strin) -> reduce((current_result, current_item) -> ((current_result === nothing) ? get(an_object, current_item, nothing) : get(current_result, current_item, nothing)), split(object_properties_string, "."); init=nothing)

JSON_OBJECT = Dict(
    "foo" => Dict(
        "bar" => "baz"
    )
)
println("JSON_OBJECT: ", JSON.json(JSON_OBJECT, 4))

println("# using JavaScript-like Optional Chaining Operator (?.) function \"optional_chaining_v1\"")

println("JSON_OBJECT?.foo?.bar: ", optional_chaining_v1(JSON_OBJECT, "foo", "bar"))
# JSON_OBJECT?.foo?.bar: baz

println("JSON_OBJECT?.foo?.baz: ", optional_chaining_v1(JSON_OBJECT, "foo", "baz"))
# JSON_OBJECT?.foo?.baz: nothing

println("# using JavaScript-like Optional Chaining Operator (?.) function \"optional_chaining_v2\"")

println("JSON_OBJECT?.foo?.bar: ", optional_chaining_v2(JSON_OBJECT, "foo.bar"))
# JSON_OBJECT?.foo?.bar: baz

println("JSON_OBJECT?.foo?.baz: ", optional_chaining_v2(JSON_OBJECT, "foo.baz"))
# JSON_OBJECT?.foo?.baz: nothing

File: codes/r/operator_optional_chaining.r
library(jsonlite)

cat("\n# JavaScript-like Optional Chaining Operator (?.) in R\n")

prettyJsonStringify <- function(anything) (if (is.null(anything) == TRUE) "NULL" else if (is.character(anything) == TRUE) paste(sep = "", "\"", anything, "\"") else prettify(toJSON(anything, pretty = TRUE, auto_unbox = TRUE), indent = 4))

# There's no JavaScript-like Optional Chaining Operator (?.) in R.
# But, we can use R built-in tryCatch function to mimic it in R.

JSON_OBJECT <- list(
    foo = list(
        bar = "baz"
    ),
    fruits = list("apple", "mango", "banana")
)
cat(paste(sep = "", "JSON_OBJECT: ", prettyJsonStringify(JSON_OBJECT), "\n"))

cat(paste(sep = "", "JSON_OBJECT?.foo?.bar: ", prettyJsonStringify(tryCatch(JSON_OBJECT$foo$bar, error = function(err) NULL)), "\n"))
# JSON_OBJECT?.foo?.bar: "baz"

cat(paste(sep = "", "JSON_OBJECT?.foo?.baz: ", prettyJsonStringify(tryCatch(JSON_OBJECT$foo$baz, error = function(err) NULL)), "\n"))
# JSON_OBJECT?.foo?.baz: NULL

cat(paste(sep = "", "JSON_OBJECT?.fruits?.[2]: ", prettyJsonStringify(tryCatch(JSON_OBJECT$fruits[[2]], error = function(err) NULL)), "\n"))
# JSON_OBJECT?.fruits?.[2]: "mango"

cat(paste(sep = "", "JSON_OBJECT?.fruits?.[5]: ", prettyJsonStringify(tryCatch(JSON_OBJECT$fruits[[5]], error = function(err) NULL)), "\n"))
# JSON_OBJECT?.fruits?.[5]: NULL
